### ISSUE TITLE
tokenise and trailing whitespace

### ISSUE THREAD

**Issue by monks1975** [spencermountain/compromise#821]
Can the compromise sentence tokeniser allow for keeping whitespace after the punctuation character? I tried a library called sbd that allowed this but it's tokenise algorithm wasn't as good. I'm trying to keep the tokenised output synchronised as an overlay with an input textarea underneath. I can always go back to regex, but If anyone has any workarounds for compromise, please give me some tips! Thanks.

**Comment by spencermountain** [comment]
ooh, that sounds like a cool project. I'd love to hear more about it.
yeah, I would do this:
```js
let doc = nlp.tokenize('foo    bar !  - baz.   --foo--   ')
//get a flat array of terms
let terms= doc.terms().json()
terms = terms.map(o => o.terms[0])
/*
[
  { text: 'foo', tags: [], pre: '', post: '    ' },
  { text: 'bar', tags: [], pre: '', post: ' !  ' },
  { text: 'baz', tags: [], pre: '- ', post: '.   ' },
  { text: 'foo', tags: [], pre: '--', post: '--   ' }
]
*/
```
There is still some opinionated lumping-splitting, but with that, you can see the whitespace on either side of each term.

If you wanted to map them back to an original string, you can do `.json({ offset: true })`

let me know if that helps
cheers

**Comment by monks1975** [comment]
Thanks @spencermountain. Unfortunately I used the tokenise only build inside a worker on the client side due to size requirements ... sounds like the terms object could be used to build quite advanced functionality though.

In the end I did some sanitising/normalising of user input with regex w/regard to spaces before tokenisation.

My project involves the classification of text for education, so Compromise is a great utility to support that.
